lib/src/main/java/com/cloudbees/groovy/cps/impl/ContinuationGroup.java

```diff
@@ -53,18 +68,51 @@ protected Next methodCall(final Env e, final SourceLocation loc, final Continuat
                 Super s = (Super) receiver;
                 v = inv.superCall(s.senderType, s.receiver, methodName, args);
             } else {
+                if (effectiveReceiver instanceof Script) {
+                    if (methodName.equals("evaluate")) { // Script.evaluate → GroovyShell.evaluate → Script.run
+                        expectedMethodNames.add("run");
+                    }
+                    // CpsScript.invokeMethod e.g. on a UserDefinedGlobalVariable cannot be predicted from here.
+                    expectedMethodNames.add("call");
+                    laxCall = !((Script) effectiveReceiver).getBinding().getVariables().containsKey(methodName); // lax unless like invokePropertyOrMissing
+                } else if (effectiveReceiver instanceof CpsBooleanClosureWrapper && methodName.equals("callForMap")) {
+                    expectedMethodNames.add("call");
+                } else if ((effectiveReceiver instanceof ListWithDefault || effectiveReceiver instanceof MapWithDefault) && methodName.equals("get")) {
+                    expectedMethodNames.add("call");
+                }
                 // TODO: spread
                 v = inv.methodCall(receiver, methodName, args);
             }
             // if this was a normal function, the method had just executed synchronously
             return k.receive(v);
         } catch (CpsCallableInvocation inv) {
+            if (!methodName.startsWith("$")) { // see TODO comment in Translator w.r.t. overloadsResolved
+                if (laxCall && inv.receiver instanceof CpsClosure) {
+                    // Potential false negative from overly lax addition above.
+                    expectedMethodNames.remove("call");
+                }
+                inv.checkMismatch(effectiveReceiver, expectedMethodNames);
+            }
             return inv.invoke(e, loc, k);
         } catch (Throwable t) {
             return throwException(e, t, loc, new ReferenceStackTrace());
         }
     }
 
+    private static Object findEffectiveReceiver(Object receiver, Map<Object, Boolean> encountered) {
+        if (!(receiver instanceof CpsClosure)) {
+            return receiver;
+        }
+        if (encountered == null) {
+            encountered = new IdentityHashMap<>();
+        }
+        if (encountered.put(receiver, true) == null) {
+            return findEffectiveReceiver(((CpsClosure) receiver).getOwner(), encountered);
```

Comment:
Any idea of how many recursive calls we'd expect for most `CpsClosures`? I doubt we need to be worried about stack overflow for non-pathological cases but I'm curious if we should expect closer to 1, 10, or 100 recursive calls on average.

Reply:
Should be one call per nested block in which a method call is made. This recursion was needed to handle cases like

```groovy
timestamps {
  node {
    ansicolor {
      something.broken()
    }
  }
}
```